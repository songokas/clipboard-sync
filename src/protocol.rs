use core::str::FromStr;

use clap::ValueEnum;
use serde::{Deserialize, Serialize};

use crate::errors::CliError;

#[derive(
    Default,
    ValueEnum,
    Serialize,
    Deserialize,
    Debug,
    Clone,
    PartialEq,
    Eq,
    Hash,
    PartialOrd,
    Ord,
    Copy,
)]
#[repr(u8)]
#[serde(rename_all = "kebab-case")]
pub enum Protocol {
    #[default]
    Basic,
    Tcp,
    TcpTls,
    Quic,
}

impl TryFrom<u8> for Protocol {
    type Error = ();

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        Ok(match value {
            0 => Self::Basic,
            1 => Self::Tcp,
            2 => Self::TcpTls,
            3 => Self::Quic,
            _ => return Err(()),
        })
    }
}

impl FromStr for Protocol {
    type Err = CliError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "basic" => Protocol::Basic,
            "tcp" => Protocol::Tcp,
            "tcp-tls" => Protocol::TcpTls,
            "quic" => Protocol::Quic,
            v => {
                return Err(CliError::ArgumentError(format!(
                    "Protocol {} is not available",
                    v
                )));
            }
        })
    }
}

impl Protocol {
    #[allow(unused_variables)]
    pub fn from(protocol_opt: Option<&str>) -> Result<Protocol, CliError> {
        let protocol = match protocol_opt {
            Some(s) => s.parse()?,
            None => Protocol::Basic,
        };
        Ok(protocol)
    }
}

impl std::fmt::Display for Protocol {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Quic => write!(f, "quic"),
            Self::Basic => write!(f, "basic"),
            Self::Tcp => write!(f, "tcp"),
            Self::TcpTls => write!(f, "tcp-tls"),
        }
    }
}
