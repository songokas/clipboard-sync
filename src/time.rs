use chrono::Utc;

#[cfg(feature = "ntp")]
use crate::errors::ConnectionError;
use crate::errors::ValidationError;
#[cfg(feature = "ntp")]
use log::{debug, warn};
#[cfg(feature = "ntp")]
use rsntp::AsyncSntpClient;
#[cfg(feature = "ntp")]
use std::sync::atomic::AtomicI64;

#[cfg(feature = "ntp")]
use std::sync::atomic::Ordering;
use std::time::Duration;

#[cfg(feature = "ntp")]
static TIME_DIFF: AtomicI64 = AtomicI64::new(0);

pub fn get_time() -> u64 {
    #[cfg(feature = "ntp")]
    {
        let diff = TIME_DIFF.load(Ordering::Relaxed);
        let now = Utc::now().timestamp();
        (now + diff) as u64
    }
    #[cfg(not(feature = "ntp"))]
    return Utc::now().timestamp() as u64;
}

pub fn validate_timestamp(
    now: u64,
    timestamp: u64,
    max_expected: Duration,
) -> Result<(), ValidationError> {
    // let now = get_time();
    let difference = if now >= timestamp {
        now - timestamp
    } else {
        timestamp - now
    };
    if max_expected.as_secs() >= difference {
        Ok(())
    } else {
        Err(ValidationError::InvalidTimestamp {
            difference,
            max_expected,
        })
    }
}

#[cfg(feature = "ntp")]
pub async fn update_time_diff(
    ntp_address: String,
    cancel: tokio_util::sync::CancellationToken,
) -> crate::defaults::ExecutorResult {
    use tokio::{
        select,
        time::{self},
    };

    let mut interval = time::interval(Duration::from_secs(1800));
    let mut success_count = 0;
    loop {
        select! {
            _ = cancel.cancelled() => {
                debug!("Ntp cancelled");
                break;
            }
            _ = interval.tick() => {
                let diff = get_time_diff(&ntp_address).await;
                TIME_DIFF.fetch_add(diff, Ordering::Relaxed);
                success_count += 1;
            }
        }
    }
    Ok(("ntp", success_count))
}

#[cfg(feature = "ntp")]
async fn get_time_diff(ntp_address: &str) -> i64 {
    let real = if let Ok(t) = get_ntp_time(ntp_address).await {
        t as i64
    } else {
        return 0;
    };
    let now = Utc::now().timestamp();
    if real >= now {
        real - now
    } else {
        now - real
    }
}

#[cfg(feature = "ntp")]
async fn get_ntp_time(ntp_address: &str) -> Result<u64, ConnectionError> {
    let client = AsyncSntpClient::new();
    let result = client.synchronize(ntp_address).await.map_err(|e| {
        warn!("Failed to synchronize time: {}", e);
        ConnectionError::FailedToConnect(format!(
            "Failed to retrieve time from ntp server {} {}",
            ntp_address, e
        ))
    })?;
    let seconds = result
        .datetime()
        .unix_timestamp()
        .expect("ntp unix timestamp");
    debug!(
        "ntp time updated. server: {} utc time in seconds: {} utc local time: {}",
        ntp_address,
        seconds.as_secs(),
        Utc::now().timestamp()
    );
    Ok(seconds.as_secs())
}

#[cfg(test)]
mod tests {
    use test_data_file::test_data_file;

    use super::*;

    #[test]
    fn test_get_time() {
        assert!(get_time() > 0);
    }

    #[test_data_file(path = "tests/samples/validate_timestamp.csv")]
    #[test]
    fn test_validate_timestamp(now: u64, timestamp: u64, max_valid_secs: u64, is_ok: bool) {
        assert_eq!(
            validate_timestamp(now, timestamp, Duration::from_secs(max_valid_secs)).is_ok(),
            is_ok
        );
    }
}



