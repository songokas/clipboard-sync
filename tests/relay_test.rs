use clipboard_sync::protocol::Protocol;
use common::run_relay_command;
use insta_cmd::assert_cmd_snapshot;
use rand::{distributions::Alphanumeric, Rng};
use std::thread;
use std::time::Duration;
use test_data_file::test_data_file;

mod common;

use crate::common::run_command;

const ENCRYPTION_KEY: &str = "12345678912345678912345678912345";
const PRIVATE_KEY: &str = "33232323233323233333333333333333";
const PUBLIC_KEY: &str = "Bj3xcJXgG4kuRolMZrIbbfY1wajtjPr4ssxSqFFhaGk=";

#[test_data_file(path = "tests/samples/relay.list")]
#[test]
fn relay_protocol(protocol: Protocol, size: usize) {
    let relay_bind = "127.0.0.1:8922";

    let t1 = thread::spawn(move || {
        assert_cmd_snapshot!(
            format!("relay_{protocol}_{size}"),
            run_relay_command(
                vec![
                    "--private-key",
                    PRIVATE_KEY,
                    "--bind-address",
                    relay_bind,
                    "--protocol",
                    &protocol.to_string(),
                    "--verbosity",
                    "debug=simple",
                ],
                Duration::from_millis(6000),
                "",
                true,
                vec![],
            )
        )
    });

    let contents: String = rand::thread_rng()
        .sample_iter(&Alphanumeric)
        .take(size)
        .map(char::from)
        .collect();
    let bind_to2 = "127.0.0.1:8923";

    thread::sleep(Duration::from_millis(200));

    let t2 = thread::spawn(move || {
        assert_cmd_snapshot!(
            format!("server_{protocol}_{size}"),
            run_command(
                vec![
                    "--key",
                    ENCRYPTION_KEY,
                    "--bind-address",
                    bind_to2,
                    "--send-using-address",
                    bind_to2,
                    "--allowed-host",
                    "127.0.0.1:8922",
                    "--protocol",
                    &protocol.to_string(),
                    "--relay-host",
                    relay_bind,
                    "--relay-public-key",
                    PUBLIC_KEY,
                    "--verbosity",
                    "debug=simple",
                    "--clipboard",
                    "/dev/stdin",
                    "--heartbeat",
                    "20",
                ],
                Duration::from_millis(6000),
                &contents,
                true,
                vec![],
            )
        )
    });

    let contents: String = rand::thread_rng()
        .sample_iter(&Alphanumeric)
        .take(size)
        .map(char::from)
        .collect();
    // let expected_size = contents.len();

    let bind_to3 = "127.0.0.1:8924";

    thread::sleep(Duration::from_millis(200));

    let t3 = thread::spawn(move || {
        assert_cmd_snapshot!(
            format!("client_{protocol}_{size}"),
            run_command(
                vec![
                    "--key",
                    ENCRYPTION_KEY,
                    "--bind-address",
                    bind_to3,
                    "--send-using-address",
                    bind_to3,
                    "--allowed-host",
                    relay_bind,
                    "--protocol",
                    &protocol.to_string(),
                    "--relay-host",
                    relay_bind,
                    "--relay-public-key",
                    PUBLIC_KEY,
                    "--verbosity",
                    "debug=simple",
                    "--clipboard",
                    "/dev/stdin",
                    "--heartbeat",
                    "20",
                ],
                Duration::from_millis(6000),
                &contents,
                true,
                vec![],
            )
        )
    });

    let _ = t1.join();
    let _ = t2.join();
    let _ = t3.join();
}
