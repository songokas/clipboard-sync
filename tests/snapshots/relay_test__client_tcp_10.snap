---
source: tests/relay_test.rs
description: "Ntp server not provided\nStarting filesystem writer\nListening on local_addr=127.0.0.1:8924 protocol=tcp multicast_addr=None\nStarting writer protocol=tcp\nStarting clipboard to protocol forwarder protocol_types=[Tcp] groups=[\"default\"]\nCreate filesystem reader {\"/dev/stdin\": {\"default\"}}\nSend clipboard on startup path=/dev/stdin\nConnect local_addr=127.0.0.1:8924 remote_addr=127.0.0.1:8922 server_name=127.0.0.1\nWatching for filesystem changes path=/dev/stdin\nStarting filesystem reader\nWaiting for stream from remote_addr=127.0.0.1:8922\nTcp receive stream local_addr=127.0.0.1:8924 remote_addr=127.0.0.1:8922\nRelay header added data_size=144\nTcp send stream stream_size=212 remote_addr=127.0.0.1:8922\nUsing existing connection local_addr=127.0.0.1:8924 remote_addr=127.0.0.1:8922 server_name=127.0.0.1\nMessage sent data_size=8 network_size=220 message_type=heartbeat\nRelay header added data_size=144\nTcp send stream stream_size=214 remote_addr=127.0.0.1:8922\nMessage sent data_size=10 network_size=222 message_type=file\nReceiver connection closed: Connection expected data, but none was received\n"
info:
  program: clipboard-sync
  args:
    - "--key"
    - "12345678912345678912345678912345"
    - "--bind-address"
    - "127.0.0.1:8924"
    - "--send-using-address"
    - "127.0.0.1:8924"
    - "--allowed-host"
    - "127.0.0.1:8922"
    - "--protocol"
    - tcp
    - "--relay-host"
    - "127.0.0.1:8922"
    - "--relay-public-key"
    - Bj3xcJXgG4kuRolMZrIbbfY1wajtjPr4ssxSqFFhaGk=
    - "--verbosity"
    - debug=simple
    - "--clipboard"
    - /dev/stdin
    - "--heartbeat"
    - "20"
snapshot_kind: text
---
success: false
exit_code: -1
----- stdout -----

----- stderr -----
Connect local_addr=127.0.0.1:8924 remote_addr=127.0.0.1:8922 server_name=127.0.0.1
Create filesystem reader {"/dev/stdin": {"default"}}
Listening on local_addr=127.0.0.1:8924 protocol=tcp multicast_addr=None
Message sent data_size=10 network_size=222 message_type=file
Message sent data_size=8 network_size=220 message_type=heartbeat
Ntp server not provided
Receiver connection closed: Connection expected data, but none was received
Relay header added data_size=144
Relay header added data_size=144
Send clipboard on startup path=/dev/stdin
Starting clipboard to protocol forwarder protocol_types=[Tcp] groups=["default"]
Starting filesystem reader
Starting filesystem writer
Starting writer protocol=tcp
Tcp receive stream local_addr=127.0.0.1:8924 remote_addr=127.0.0.1:8922
Tcp send stream stream_size=212 remote_addr=127.0.0.1:8922
Tcp send stream stream_size=214 remote_addr=127.0.0.1:8922
Using existing connection local_addr=127.0.0.1:8924 remote_addr=127.0.0.1:8922 server_name=127.0.0.1
Waiting for stream from remote_addr=127.0.0.1:8922
Watching for filesystem changes path=/dev/stdin
